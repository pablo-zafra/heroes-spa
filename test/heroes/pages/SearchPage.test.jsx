const { render, screen, getByRole, fireEvent } = require("@testing-library/react");
const { MemoryRouter } = require("react-router-dom");
const { SearchPage } = require("../../../src/heroes/pages/SearchPage");

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));

describe('Pruebas en <SearchPage />', () => {

    beforeEach(() => { jest.clearAllMocks() });
    
    test('Debe mostrar valores por defecto', () => { 
        
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        // console.log( container );
        expect( container ).toMatchSnapshot();

    });
     
    test('Debe mostrar en el input el valor del query-string y el resultado de bÃºsqueda.', () => { 
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect( input.value ).toBe( 'batman' );

        const img = screen.getAllByRole('img')[0];
        expect( img.src ).toContain( 'batman' );

        // const searchAHeroBox = screen.getByText('Search a hero');
        expect( screen.queryByText('Search a hero') ).toBeFalsy();

    });

    test('Debe mostrar error si no se encuentra el hero (batman123)', () => { 
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        );

        const error = screen.queryByText( 'No hero with:' );
        expect( error ).toBeTruthy();

    });

    test('Debe de llamar el navigate a la pantalla nueva', () => { 

        const inputValue = 'superman';
        
        render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change( input, { target: { name: 'searchText', value: inputValue}});
        
        const btnSearch = screen.getByRole('button');
        fireEvent.click(btnSearch);
        
        expect( mockedUseNavigate ).toHaveBeenCalledWith(`?q=${ inputValue }`);

     })

 })