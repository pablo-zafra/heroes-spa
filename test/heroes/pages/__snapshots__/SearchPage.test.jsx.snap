// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> Debe mostrar valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row animate__animated animate__fadeInUp"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a Hero"
          type="text"
          value=""
        />
      </form>
      <button
        class="btn btn-outline-primary mt-3"
      >
        Search
      </button>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results:
      </h4>
      <hr />
      <div
        class="alert alert-primary"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
